lib/connectors/memory.js

```diff
@@ -351,28 +351,46 @@ Memory.prototype.fromDb = function(model, data) {
   data = deserialize(data);
   var props = this._models[model].properties;
   for (var key in data) {
-    var val = data[key];
-    if (val === undefined || val === null) {
-      continue;
-    }
-    if (props[key]) {
-      switch (props[key].type.name) {
-        case 'Date':
-          val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
-          break;
-        case 'Boolean':
-          val = Boolean(val);
-          break;
-        case 'Number':
-          val = Number(val);
-          break;
-      }
-    }
-    data[key] = val;
+    data[key] = this._castPropertyValue(key, data[key], props);
   }
   return data;
 };
 
+Memory.prototype._castPropertyValue = function(prop, val, props) {
+  var self = this;
+  if (val === undefined || val === null || !props[prop]) {
+    return val;
+  }
+
+  if (Array.isArray(val)) {
+    return val.map(function(val) {
+      return self._castPropertyValue(prop, val, props);
+    });
+  }
+
+  var isArray = Array.isArray(props[prop].type);
+  var propType = isArray ? props[prop].type[0] : props[prop].type;
+
+  switch (propType.name) {
+    case 'Date':
+      val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
+      break;
+    case 'Boolean':
+      val = Boolean(val);
+      break;
+    case 'Number':
+      val = Number(val);
+      break;
+    case 'ModelConstructor':
+      for (var subProp in val) {
+        val[subProp] = this._castPropertyValue(subProp, val[subProp], propType.definition.properties);
```

Comment:
I am little bit worried about this block. I think `ModelConstructor` is used only for anonymous types, but what if the property is referring to a named type?

```js
const Address = createModel('Address', { street: string, ... });
const Person = createModel('Person', {
  address: Address,
  // ...
});
```

Wouldn't the `propType.name` be `Address` for `Person.address`?

Also, are we sure that `propType.definition` and `propType.definition.properties` is always defined?